File: crates/process_web/src/api/datasource_list.rs
use std::sync::Arc;
use axum::extract::{Path, State};
use axum::{Json, Router};
use axum::routing::{get, post};
use serde::Deserialize;
use ts_rs::TS;
use crate::api::common::{AppError, AppState, PaginationPayload, ResJson, ResJsonWithPagination};
use crate::{bool_response, data_response, pagination_response};
use crate::entity::datasource_list::Model;
use crate::service::datasource_list_service::DataSourceListService;

pub fn set_routes() -> Router<Arc<AppState>> {
    let routes = Router::new()
        .route("/find_by_id/:id", get(find_by_id))
        .route("/list", post(list))
        .route("/add", post(add))
        .route("/update_by_id/:id", post(update_by_id))
        .route("/del/:id", get(del));

    routes
}

#[derive(Deserialize, TS)]
#[ts(
export,
export_to = "ui/api/models/auto-generates/DatasourceListParams.ts",
rename = "DatasourceListParams"
)]
pub struct ListParams {
    pub database_name: Option<String>,
}

async fn find_by_id(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
) -> anyhow::Result<ResJson<Model>, AppError> {
    let res = DataSourceListService::find_by_id(&state.conn, id).await;
    data_response!(res)
}

async fn list(
    state: State<Arc<AppState>>,
    Json(payload): Json<PaginationPayload<ListParams>>,
) -> anyhow::Result<ResJsonWithPagination<Model>, AppError> {
    let res = DataSourceListService::list(
        &state.conn,
        payload.current,
        payload.page_size,
        payload.data,
    )
        .await;

    pagination_response!(res, payload.current, payload.page_size)
}

async fn add(
    state: State<Arc<AppState>>,
    Json(payload): Json<Model>,
) -> anyhow::Result<ResJson<Model>, AppError> {
    let res = DataSourceListService::add(&state.conn, payload).await;

    data_response!(res)
}


async fn update_by_id(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
    Json(payload): Json<Model>,
) -> anyhow::Result<ResJson<Model>, AppError> {
    let res = DataSourceListService::update_by_id(&state.conn, id, payload).await;

    data_response!(res)
}

async fn del(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
) -> anyhow::Result<ResJson<bool>, AppError> {
    let res = DataSourceListService::delete(&state.conn, id).await;

    bool_response!(res)
}

File: crates/process_web/ui/api/data_source_list.ts
import { DataSourceList } from "@/api/models/DataSourceList";
import {
  http_get,
  http_post,
  PaginationPayload,
  ResJson,
  ResJsonWithPagination,
} from "@/api/common";
import { DataSourceListParams } from "@/api/models/DataSourceListParams";
import { QueryTableColumnsParameters } from "@/api/models/QueryTableColumnsParameters";

export const PREFIX = "/api/data_source_list";

export const FIND_BY_ID = `${PREFIX}/find_by_id/`;

export async function find_by_id(id: number): Promise<ResJson<DataSourceList>> {
  return http_get(`${FIND_BY_ID}${id}`);
}

export const LIST = `${PREFIX}/list/`;
export async function list(
  payload: PaginationPayload<DataSourceListParams>
): Promise<ResJsonWithPagination<DataSourceList>> {
  return http_post(`${PREFIX}/list`, {
    body: JSON.stringify(payload),
  });
}

export const ADD = `${PREFIX}/add/`;
export async function add(
  payload: DataSourceList
): Promise<ResJson<DataSourceList>> {
  console.log(payload);
  return http_post(`${PREFIX}/add`, {
    body: JSON.stringify(payload),
  });
}

export const UPDATE_BY_ID = `${PREFIX}/update_by_id/`;
export async function update_by_id(
  id: number,
  payload: DataSourceList
): Promise<ResJson<DataSourceList>> {
  return http_post(`${PREFIX}/update_by_id/${id}`, {
    body: JSON.stringify(payload),
  });
}

export const DEL = `${PREFIX}/del/`;
export async function del(id: number): Promise<ResJson<boolean>> {
  return http_get(`${PREFIX}/del/${id}`);
}

export const QUERY_TABLE_COLUMNS = `${PREFIX}/query_table_columns/`;
export async function query_table_columns(
  payload: QueryTableColumnsParameters
): Promise<ResJson<Record<string, string>[]>> {
  return http_post(`${PREFIX}/query_table_columns`, {
    body: JSON.stringify(payload),
  });
}
